### PR TITLE
Fix cos flow

### PR DESCRIPTION
**Pull request by Prajyot-Parab**
**What this PR does / why we need it**:
- Fix cos flow

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
Fix cos flow
```

### CODE REVIEW

File: cloud/scope/powervs_cluster.go [hunk 1]
```diff
@@ -1453,14 +1449,10 @@ func (s *PowerVSClusterScope) ReconcileCOSInstance() error {
...
-	region := s.IBMPowerVSCluster.Spec.CosInstance.BucketRegion
-	// if the bucket region is not set, use vpc region
-	if region == "" {
-		vpcDetails := s.VPC()
-		if vpcDetails == nil || vpcDetails.Region == nil {
-			return fmt.Errorf("failed to determine cos bucket region, both buckeet region and vpc region not set")
-		}
-		region = *vpcDetails.Region
+
+	region, regerr := s.bucketRegion()
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster.go:1453]:
```suggestion
	region, err := s.bucketRegion()
```

Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster.go:1453]:
```suggestion
	region, err := s.bucketRegion()
```

File: cloud/scope/powervs_cluster.go [hunk 2]
```diff
@@ -1977,3 +1969,26 @@ func (s *PowerVSClusterScope) isResourceCreatedByController(resourceType infrav1
 	}
 	return false
 }
+
+func (s *PowerVSClusterScope) bucketName() string {
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster.go:1973]: can't this be moved under GetServiceName() method? we can create a resource type as bucket .

File: cloud/scope/powervs_cluster.go [hunk 3]
```diff
@@ -1977,3 +1969,26 @@ func (s *PowerVSClusterScope) isResourceCreatedByController(resourceType infrav1
...
+	if s.COSInstance() == nil || s.COSInstance().BucketName == "" {
+		return fmt.Sprintf("%s-%s", *s.GetServiceName(infrav1beta2.ResourceTypeCOSInstance), "bucket")
+	}
+	return s.COSInstance().BucketName
+}
+
+func (s *PowerVSClusterScope) bucketRegion() (string, error) {
+	var region string
+	if s.COSInstance() != nil && s.COSInstance().BucketRegion != "" {
+		region = s.COSInstance().BucketRegion
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster.go:1983]: you can simply return region from here

File: cloud/scope/powervs_cluster.go (hunk 2 above)
Comment (mkumatag) [inline cloud/scope/powervs_cluster.go:1973]: I see same functions in the machinescope as well, lets add a todo to get rid off these duplicate functions

File: cloud/scope/powervs_cluster.go [hunk 4]
```diff
@@ -1977,3 +1969,26 @@ func (s *PowerVSClusterScope) isResourceCreatedByController(resourceType infrav1
...
+	}
+	// if the bucket region is not set, use vpc region
+	if region == "" {
+		vpcDetails := s.VPC()
+		if vpcDetails == nil || vpcDetails.Region == nil {
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster.go:1988]:
```suggestion
		if vpcDetails != nil && vpcDetails.Region != nil {
		  return *vpcDetails.Region
		}
```

File: cloud/scope/powervs_machine.go [hunk 5]
```diff
@@ -1074,3 +1075,26 @@ func (m *PowerVSMachineScope) APIServerPort() int32 {
 	}
 	return infrav1beta2.DefaultAPIServerPort
 }
+
+func (m *PowerVSMachineScope) bucketName() string {
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_machine.go:1079]: I know its a code duplication, lets add a todo to handle it later, I think it needs some refactoring

**Review by Karthik-K-N — COMMENTED** [review]
Couple of suggestions

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: cloud/scope/powervs_cluster.go (hunk 2 above)
Comment (Prajyot-Parab) [reply cloud/scope/powervs_cluster.go]: Done.

File: cloud/scope/powervs_machine.go (hunk 5 above)
Comment (Prajyot-Parab) [reply cloud/scope/powervs_machine.go:1081]: Done.

File: cloud/scope/powervs_cluster.go (hunk 4 above)
Comment (Prajyot-Parab) [reply cloud/scope/powervs_cluster.go]: Done.

File: cloud/scope/powervs_cluster.go (hunk 2 above)
Comment (Prajyot-Parab) [reply cloud/scope/powervs_cluster.go]: Done.

File: cloud/scope/powervs_cluster.go (hunk 1 above)
Comment (Prajyot-Parab) [reply cloud/scope/powervs_cluster.go]: Done.

File: cloud/scope/powervs_machine.go [hunk 6]
```diff
@@ -589,14 +596,9 @@ func (m *PowerVSMachineScope) createCOSClient() (*cos.Service, error) {
...
-	region := m.IBMPowerVSCluster.Spec.CosInstance.BucketRegion
-	// if the bucket region is not set, use vpc region
+	region := m.bucketRegion()
 	if region == "" {
-		vpcDetails := m.IBMPowerVSCluster.Spec.VPC
-		if vpcDetails == nil || vpcDetails.Region == nil {
-			return nil, fmt.Errorf("failed to determine cos bucket region, both buckeet region and vpc region not set")
-		}
-		region = *vpcDetails.Region
+		return nil, fmt.Errorf("failed to determine cos bucket region, both buckeet region and vpc region not set")
```
Comment (Amulyam24) [inline cloud/scope/powervs_machine.go:601]:
```suggestion
		return nil, fmt.Errorf("failed to determine cos bucket region, both bucket region and vpc region not set")
```

File: cloud/scope/powervs_machine.go [hunk 7]
```diff
@@ -408,7 +408,12 @@ func (m *PowerVSMachineScope) createIgnitionData(data []byte) (string, error) {
 	key := m.bootstrapDataKey()
 	m.Info("bootstrap data key", "key", key)
 
-	bucket := m.IBMPowerVSCluster.Spec.CosInstance.BucketName
+	bucket := m.bucketName()
+	region := m.bucketRegion()
+	if region == "" {
+		return "", fmt.Errorf("failed to determine cos bucket region, both buckeet region and vpc region not set")
```
Comment (Amulyam24) [inline cloud/scope/powervs_machine.go:414]:
```suggestion
		return "", fmt.Errorf("failed to determine cos bucket region, both bucket region and vpc region not set")
```

**Review by Amulyam24 — COMMENTED** [review]
minor nit, otherwise LGTM!

File: cloud/scope/powervs_cluster.go [hunk 8]
```diff
@@ -1977,3 +1974,19 @@ func (s *PowerVSClusterScope) isResourceCreatedByController(resourceType infrav1
...
 }
+
+// TODO: duplicate function, optimize it.
+func (s *PowerVSClusterScope) bucketRegion() string {
+	var region string
+	if s.COSInstance() != nil && s.COSInstance().BucketRegion != "" {
+		return s.COSInstance().BucketRegion
+	}
+	// if the bucket region is not set, use vpc region
+	if region == "" {
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster.go:1985]:
```suggestion
	if region == "" {
```

File: cloud/scope/powervs_cluster.go [hunk 9]
```diff
@@ -1977,3 +1974,19 @@ func (s *PowerVSClusterScope) isResourceCreatedByController(resourceType infrav1
...
+		vpcDetails := s.VPC()
+		if vpcDetails != nil && vpcDetails.Region != nil {
+			return *vpcDetails.Region
+		}
+	}
+	return region
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster.go:1991]:
```suggestion
	// if the bucket region is not set, use vpc region
	vpcDetails := s.VPC()
	if vpcDetails != nil && vpcDetails.Region != nil {
		return *vpcDetails.Region
	}
	return ""
```

**Review by Karthik-K-N — COMMENTED** [review]
Small suggestion otherwise LGTM

**Review by Karthik-K-N — COMMENTED** [review]
/lgtm

File: api/v1beta2/ibmpowervscluster_webhook.go [hunk 10]
```diff
@@ -137,19 +137,5 @@ func (r *IBMPowerVSCluster) validateIBMPowerVSClusterCreateInfraPrereq() *field.
 		return nil
```
Comment (Karthik-K-N) [inline api/v1beta2/ibmpowervscluster_webhook.go:137]: Please remove this check as well, Its no more needed.

Comment (Prajyot-Parab) [reply api/v1beta2/ibmpowervscluster_webhook.go]: Done.

File: api/v1beta2/ibmpowervscluster_webhook.go [hunk 11]
```diff
@@ -96,60 +90,3 @@ func (r *IBMPowerVSCluster) validateIBMPowerVSClusterNetwork() *field.Error {
...
-		return field.Invalid(field.NewPath("spec.vpc.region"), r.Spec.VPC.Region, "value of VPC region is empty")
-	}
-
-	if r.Spec.ResourceGroup == nil {
-		return field.Invalid(field.NewPath("spec.resourceGroup"), r.Spec.ResourceGroup, "value of resource group is empty")
-	}
-
-	if r.Spec.Ignition == nil {
-		return nil
-	}
```
Comment (Karthik-K-N) [inline api/v1beta2/ibmpowervscluster_webhook.go:138]: Hey, You are supposed remove it from here, Not the whole file, We still need webhook for verifying zone vpc region.

**Review by Karthik-K-N — COMMENTED** [review]
/lgtm

**Pull request merged**